Add option to get hoplimit from ancillary data using recvmsg()

Cristib05 opened the pull request:
This PR adds support for retrieving hop limit/ttl values from ancillary data when using recvmsg() function.

Relates to #94156

Cristib05 committed:
net: context: Add support for setting receive hop limit option

Add support for setting IPV6_RECVHOPLIMIT or IP_RECVTTL socket option.

Signed-off-by: Cristian Bulacu <cristian.bulacu@nxp.com>

Cristib05 committed:
net: socket: Add support for setting hop limit options

Add IP_RECVTTL or IPV6_RECVHOPLIMIT BSD socket options
that can be used to extract ttl/hop limit from ancillary data
in recvmsg() call.

Signed-off-by: Cristian Bulacu <cristian.bulacu@nxp.com>

Cristib05 commented:
Hello @rlubos @jukkar.
In this PR I have added support for retrieving hop limit/ttl from ancillary data when using recvmsg(). It is somehow an addition to #65694.

PR is marked as draft because I need to implement tests for this new feature. I was thinking maybe I can reuse `run_ancillary_recvmsg_test` function and make it more generic, to cover more ancillary options, not only pktinfo. Thoughts on this?  

I do have some questions regarding to `add_pktinfo` function from `sockets_inet.c`. Why is `insert_pktinfo()` function called with `IPV6_RECVPKTINFO` parameter? From RFC3542, I have seen that `IPV6_PKTINFO` is stated as `optname/cmsg_type.` I have seen that for BSD `setsockopt` might be called with `IPV6_PKTINFO`, but I haven't seen `IPV6_RECVPKTINFO` as `cmsg_type`.
Regarding the test for pktinfo functionality, I have also seen this check `cmsg->cmsg_type == IPV6_RECVPKTINFO`. I am not very familiar with socket programming, if I said something wrong, please let me know.

Another thing that I saw in tests, is that `run_ancillary_recvmsg_test` is called for both IPV4 and IPV6 scenarios. When `server_sock` is populated using `prepare_sock_udp_v6`, is it valid to call `rv = zsock_setsockopt(server_sock, IPPROTO_IP, IP_PKTINFO, &opt, optlen)`? 

Thanks!

jukkar commented on include/zephyr/net/net_context.h:
```diff
@@ -370,6 +370,10 @@ __net_socket struct net_context {
 		bool recv_pktinfo;
 #endif
+#if defined(CONFIG_NET_CONTEXT_RECV_HOPLIMIT)
+		/** Receive hop limit as ancillary data in recvmsg() call */
```
```suggestion
		/** Receive IPv6 hop limit or IPv4 TTL as ancillary data in recvmsg() call */
```

jukkar commented on subsys/net/ip/Kconfig:
```diff
@@ -608,6 +608,14 @@ config NET_CONTEXT_RECV_PKTINFO
 	  socket.
 
+config NET_CONTEXT_RECV_HOPLIMIT
+	bool "Add receive PKTINFO support to net_context"
```
pktinfo looks wrong here (copy/paste error?)

jukkar commented on subsys/net/ip/Kconfig:
```diff
@@ -608,6 +608,14 @@ config NET_CONTEXT_RECV_PKTINFO
 
+config NET_CONTEXT_RECV_HOPLIMIT
+	bool "Add receive PKTINFO support to net_context"
+	depends on NET_UDP
```
The option should work for TCP sockets for IPv6, at least that is mentioned here https://www.ibm.com/docs/en/zos/2.4.0?topic=soadsiiil-options-that-provide-information-about-packets-that-have-been-received
although I did not spot mention of TCP in the RFC https://www.ietf.org/rfc/rfc3542.txt
But the `IP_RECVTTL` specifically mentions that stream socket is not supported.
So we could try to reflect this in the code by adding needed runtime checks there.

rlubos replied:
I'd personally focus on datagrams only for now. That use case for stream sockets seems a bit odd.

jukkar commented on subsys/net/lib/sockets/sockets_inet.c:
```diff
@@ -2584,6 +2653,23 @@ int zsock_setsockopt_ctx(struct net_context *ctx, int level, int optname,
 
+		case IP_RECVTTL:
+			if (IS_ENABLED(CONFIG_NET_IPV4) &&
+				IS_ENABLED(NET_CONTEXT_RECV_HOPLIMIT)) {
```
Please align the `IS_ENABLED` macros (could be also that GH renders this incorrectly)

jukkar commented on subsys/net/lib/sockets/sockets_inet.c:
```diff
@@ -2584,6 +2653,23 @@ int zsock_setsockopt_ctx(struct net_context *ctx, int level, int optname,
+			if (IS_ENABLED(CONFIG_NET_IPV4) &&
+				IS_ENABLED(NET_CONTEXT_RECV_HOPLIMIT)) {
+				ret = net_context_set_option(ctx,
+								 NET_OPT_RECV_HOPLIMIT,
+								 optval,
+								 optlen);
```
Alignment might be wrong here (could be also that the web page just shows this incorrectly)

jukkar replied:
There is definitely something fishy with the alignment. The code should look like this with proper indentation
```
	case IP_RECVTTL:
		if (IS_ENABLED(CONFIG_NET_IPV4) &&
		    IS_ENABLED(CONFIG_NET_CONTEXT_RECV_HOPLIMIT)) {
			ret = net_context_set_option(ctx,
						     NET_OPT_RECV_HOPLIMIT,
						     optval,
						     optlen);
```
It looked like there are too many tabs used, have you set tab size to 8 in your editor.

jukkar commented on subsys/net/lib/sockets/sockets_inet.c:
```diff
@@ -2704,6 +2790,23 @@ int zsock_setsockopt_ctx(struct net_context *ctx, int level, int optname,
 
+		case IPV6_RECVHOPLIMIT:
+			if (IS_ENABLED(CONFIG_NET_IPV6) &&
+				IS_ENABLED(CONFIG_NET_CONTEXT_RECV_HOPLIMIT)) {
```
same here

jukkar commented on subsys/net/lib/sockets/sockets_inet.c:
```diff
@@ -2704,6 +2790,23 @@ int zsock_setsockopt_ctx(struct net_context *ctx, int level, int optname,
+			if (IS_ENABLED(CONFIG_NET_IPV6) &&
+				IS_ENABLED(CONFIG_NET_CONTEXT_RECV_HOPLIMIT)) {
+				ret = net_context_set_option(ctx,
+								 NET_OPT_RECV_HOPLIMIT,
+								 optval,
+								 optlen);
```
and here

jukkar commented on subsys/net/ip/Kconfig:
```diff
@@ -608,6 +608,14 @@ config NET_CONTEXT_RECV_PKTINFO
+	bool "Add receive PKTINFO support to net_context"
+	depends on NET_UDP
+	help
+	  Allow to set the IPV6_RECVHOPLIMIT flags on a socket.
```
Should we mention `IP_RECVTTL` here too?

jukkar commented on include/zephyr/net/net_context.h:
```diff
@@ -370,6 +370,10 @@ __net_socket struct net_context {
 		/** Receive network packet information in recvmsg() call */
```
Commit message of `net: context: Add support for setting receive hoplimit option` should mention that this also supports IPv4 `IP_RECVTTL` option.

Cristib05 commented:
> > PR is marked as draft because I need to implement tests for this new feature. I was thinking maybe I can reuse `run_ancillary_recvmsg_test` function and make it more generic, to cover more ancillary options, not only pktinfo. Thoughts on this?
> 
> The function looks quite complex already, it might be easier to create a separate test function for these options.
> 
> > I do have some questions regarding to add_pktinfo function from sockets_inet.c. Why is insert_pktinfo() function called with IPV6_RECVPKTINFO parameter? From RFC3542, I have seen that IPV6_PKTINFO is stated as optname/cmsg_type. I have seen that for BSD setsockopt might be called with IPV6_PKTINFO, but I haven't seen IPV6_RECVPKTINFO as cmsg_type.
> > Regarding the test for pktinfo functionality, I have also seen this check cmsg->cmsg_type == IPV6_RECVPKTINFO. I am not very familiar with socket programming, if I said something wrong, please let me know.
> 
> Indeed, the cmsg_type should be `IPV6_PKTINFO` in this case. Would you be able to send a fix to that?
> 
> > Another thing that I saw in tests, is that run_ancillary_recvmsg_test is called for both IPV4 and IPV6 scenarios. When server_sock is populated using prepare_sock_udp_v6, is it valid to call rv = zsock_setsockopt(server_sock, IPPROTO_IP, IP_PKTINFO, &opt, optlen)?
> 
> That looks like a bug too, the setsockopt should return an error if trying to set IPv4 option to IPv6 socket. Not sure if that is also true if IPv6 is also serving IPv4 data (when V6_ONLY option is not set).

I will then create a new testcast, just for hoplimit.
I will create a new commit in this PR to address `IPV6_PKTINFO` issue and I will try to see what happens for the `zsock_setsockopt`

Cristib05 committed:
net: socket: Fix recvmsg() pktinfo support.

Fix an issue where IPV6_RECVPKTINFO was used instead of IPV6_PKTINFO
when insert_pktinfo was called.

Signed-off-by: Cristian Bulacu <cristian.bulacu@nxp.com>

Cristib05 committed:
tests: net: socket: Fix recv pktinfo ancillary test for recvmsg()

Fixed setsockopt based on socket family and added correct cmsg_type for
parsing ancillary data.

Signed-off-by: Cristian Bulacu <cristian.bulacu@nxp.com>

Cristib05 committed:
tests: net: socket: Add test for hop limit ancillary data using recvmsg()

Testing that recvmsg() returns IPv4 TTL and IPv6 hop limit ancillary
data.

Signed-off-by: Cristian Bulacu <cristian.bulacu@nxp.com>

rlubos commented on subsys/net/lib/sockets/sockets_inet.c:
```diff
@@ -1061,6 +1061,68 @@ static int add_pktinfo(struct net_context *ctx,
+	net_pkt_cursor_init(pkt);
+
+	if (IS_ENABLED(CONFIG_NET_IPV4) && net_pkt_family(pkt) == AF_INET) {
+		NET_PKT_DATA_ACCESS_CONTIGUOUS_DEFINE(ipv4_access,
```
When processing IPv4 the stack sets the TTL value in the net_pkt, I think it would be way easier to read it with `net_pkt_ipv4_ttl()` instead fetching the ip header again.

rlubos commented on subsys/net/lib/sockets/sockets_inet.c:
```diff
@@ -1061,6 +1061,68 @@ static int add_pktinfo(struct net_context *ctx,
+	}
+
+	if (IS_ENABLED(CONFIG_NET_IPV6) && net_pkt_family(pkt) == AF_INET6) {
+		NET_PKT_DATA_ACCESS_CONTIGUOUS_DEFINE(ipv6_access,
```
Ditto, hop limit can be read from the packet with `net_pkt_ipv6_hop_limit()`.

Cristib05 commented on tests/net/socket/udp/src/main.c:
```diff
@@ -3116,6 +3116,197 @@ ZTEST(net_socket_udp, test_40_clamp_udp_tcp_port_range)
+
+	cmsg = CMSG_FIRSTHDR(&msg);
+	cmsg->cmsg_len = CMSG_LEN(sizeof(int));
+	if (client_addr->sa_family == AF_INET) {
+		cmsg->cmsg_level = IPPROTO_IPV6;
+	} else {
+		cmsg->cmsg_level = IPPROTO_IP;
+	}
```
> Isn't `IPPROTO_IPV6`/`IPPROTO_IP` mixed up here (I figure for `AF_INET` we should set `IPPROTO_IP` level)?

it is, my mistake

Cristib05 commented:
> BTW, no need to keep this as a draft any longer, the PR looks pretty ok as a normal reviewable PR.

I will address the latest feedback and add copyrigths, then I will mark it as ready for review.

Cristib05 committed:
net: context: Add support for setting hop limit from ancillary data

Add support to parse over msghdr and set pkt hop limit, if needed, when
sendmsg() is used.

Signed-off-by: Cristian Bulacu <cristian.bulacu@nxp.com>

Cristib05 commented on tests/net/socket/udp/src/main.c:
```diff
@@ -3116,6 +3116,197 @@ ZTEST(net_socket_udp, test_40_clamp_udp_tcp_port_range)
 #endif
 }
 
+static void run_ancillary_recvmsg_hoplimit_test(int client_sock,
+				       struct sockaddr *client_addr,
+				       int client_addr_len,
+				       int server_sock,
+				       struct sockaddr *server_addr,
+				       int server_addr_len)
```
> nit, arguments alignment.

Can you please let me know the rule on the arguments alignment? I looked over the documentation but I haven't seen something like this, or I have completely missed.

rlubos replied:
Not sure where is this documented, but generally we eitehr align arguments to the opening bracket, or move all arguments to the line below (then use whatever alingment that fits, just be consistent if you have arguments in several lines)

jukkar commented on subsys/net/lib/sockets/sockets_inet.c:
```diff
@@ -1049,7 +1049,7 @@ static int add_pktinfo(struct net_context *ctx,
 		info.ipi6_ifindex = ctx->iface;
 
-		ret = insert_pktinfo(msg, IPPROTO_IPV6, IPV6_RECVPKTINFO,
+		ret = insert_pktinfo(msg, IPPROTO_IPV6, IPV6_PKTINFO,
```
@Cristib05 please do not mark conversations as resolved but let the reviewer do that after he has verified that the change is done

Cristib05 replied:
@jukkar, understood.

jukkar replied:
No worries, there was not rules earlier but atm the "Resolve conversation" button usage is described in the documentation
https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#workflow-suggestions-that-help-reviewers

jukkar commented on tests/net/socket/udp/src/main.c:
```diff
@@ -3214,6 +3215,268 @@ ZTEST(net_socket_udp, test_42_v6_dgram_peer_addr_reset)
 }
 
+static void comm_sendmsg_recvmsg_hop_limit(int client_sock,
+				 struct sockaddr *client_addr,
+				 socklen_t client_addrlen,
+				 const struct msghdr *client_msg,
+				 int server_sock,
+				 struct sockaddr *server_addr,
+				 socklen_t server_addrlen,
+				 struct msghdr *msg,
+				 void *cmsgbuf, int cmsgbuf_len,
+				 bool expect_control_data)
```
Indentation is off here (one tab and two spaces are needed)

jukkar commented on tests/net/socket/udp/src/main.c:
```diff
@@ -3214,6 +3215,268 @@ ZTEST(net_socket_udp, test_42_v6_dgram_peer_addr_reset)
+	*(int *)CMSG_DATA(cmsg) = send_hop_limit;
+
+	comm_sendmsg_recvmsg_hop_limit(client_sock,
+			     client_addr,
+			     client_addr_len,
+			     &msg,
+			     server_sock,
+			     server_addr,
+			     server_addr_len,
+			     &server_msg,
+			     &cmsgbuf.buf,
+			     sizeof(cmsgbuf.buf),
+			     true);
```
indentation

jukkar commented on tests/net/socket/udp/src/main.c:
```diff
@@ -3116,6 +3116,197 @@ ZTEST(net_socket_udp, test_40_clamp_udp_tcp_port_range)
+	     prevcmsg = cmsg, cmsg = CMSG_NXTHDR(&server_msg, cmsg)) {
+		if (client_addr->sa_family == AF_INET) {
+			if (cmsg->cmsg_level == IPPROTO_IP &&
+				cmsg->cmsg_type == IP_TTL) {
```
alignment is wrong, please do like this
```

			if (cmsg->cmsg_level == IPPROTO_IP &&
			    cmsg->cmsg_type == IP_TTL) {
```

jukkar commented on tests/net/socket/udp/src/main.c:
```diff
@@ -3214,6 +3215,268 @@ ZTEST(net_socket_udp, test_42_v6_dgram_peer_addr_reset)
+	*(int *)CMSG_DATA(cmsg) = send_hop_limit;
+
+	comm_sendmsg_recvmsg_hop_limit(client_sock,
+			     client_addr,
+			     client_addr_len,
+			     &msg,
+			     server_sock,
+			     server_addr,
+			     server_addr_len,
+			     &server_msg,
+			     &cmsgbuf.buf,
+			     sizeof(cmsgbuf.buf),
+			     true);
```
alignment

jukkar commented on tests/net/socket/udp/src/main.c:
```diff
@@ -3214,6 +3215,268 @@ ZTEST(net_socket_udp, test_42_v6_dgram_peer_addr_reset)
+	     prevcmsg = cmsg, cmsg = CMSG_NXTHDR(&server_msg, cmsg)) {
+		if (client_addr->sa_family == AF_INET) {
+			if (cmsg->cmsg_level == IPPROTO_IP &&
+				cmsg->cmsg_type == IP_TTL) {
```
ditto

jukkar commented on tests/net/socket/udp/src/main.c:
```diff
@@ -3214,6 +3215,268 @@ ZTEST(net_socket_udp, test_42_v6_dgram_peer_addr_reset)
+			}
+		} else {
+			if (cmsg->cmsg_level == IPPROTO_IPV6 &&
+				cmsg->cmsg_type == IPV6_HOPLIMIT) {
```
ditto

jukkar commented on tests/net/socket/udp/src/main.c:
```diff
@@ -3214,6 +3215,268 @@ ZTEST(net_socket_udp, test_42_v6_dgram_peer_addr_reset)
+	prepare_sock_udp_v4(MY_IPV4_ADDR, SERVER_PORT, &server_sock, &server_addr);
+
+	run_ancillary_recvmsg_hoplimit_test(client_sock,
+				   (struct sockaddr *)&client_addr,
```
ditto

jukkar commented on tests/net/socket/udp/src/main.c:
```diff
@@ -3214,6 +3215,268 @@ ZTEST(net_socket_udp, test_42_v6_dgram_peer_addr_reset)
+	prepare_sock_udp_v6(MY_IPV6_ADDR, SERVER_PORT, &server_sock, &server_addr);
+
+	run_ancillary_recvmsg_hoplimit_test(client_sock,
+				   (struct sockaddr *)&client_addr,
```
ditto
